Library/OcDeviceMiscLib/SetResizableBar.c

```diff
@@ -309,8 +381,234 @@ SetResizableBarOnDevice (
   return EFI_SUCCESS;
 }
 
+STATIC
 EFI_STATUS
-ResizeGpuBars (
+SetResizableBarOnDeviceRbIo (
+  IN EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL  *PciRootBridgeIo,
+  IN UINT64                           PciAddress,
+  IN PCI_BAR_SIZE                     Size,
+  IN BOOLEAN                          Increase
+  )
+{
+  EFI_STATUS                                               Status;
+  UINT32                                                   ResizableBarOffset;
+  PCI_EXPRESS_EXTENDED_CAPABILITIES_RESIZABLE_BAR_ENTRY    Entries[PCI_MAX_BAR];
+  PCI_EXPRESS_EXTENDED_CAPABILITIES_RESIZABLE_BAR_CONTROL  ResizableBarControl;
+  UINT32                                                   Offset;
+  UINT32                                                   Index;
+  UINT32                                                   ResizableBarNumber;
+  UINT64                                                   Capabilities;
+  UINT64                                                   NewCapabilities;
+  UINT32                                                   OldBar[PCI_MAX_BAR];
+  UINT32                                                   NewBar[PCI_MAX_BAR];
+  INTN                                                     Bit;
+
+  Status = LocatePciCapabilityRbIo (
+             PciRootBridgeIo,
+             PciAddress,
+             PCI_EXPRESS_EXTENDED_CAPABILITY_RESIZABLE_BAR_ID,
+             &ResizableBarOffset
+             );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_INFO, "OCDM: RBAR is unsupported by device - %r\n", Status));
+    return EFI_UNSUPPORTED;
+  }
+
+  ResizableBarControl.Uint32 = 0;
+  Offset                     = ResizableBarOffset + sizeof (PCI_EXPRESS_EXTENDED_CAPABILITIES_HEADER)
+                               + sizeof (PCI_EXPRESS_EXTENDED_CAPABILITIES_RESIZABLE_BAR_CAPABILITY);
+  Status = PciRootBridgeIo->Pci.Read (
+                                  PciRootBridgeIo,
+                                  EfiPciWidthUint8,
+                                  PciAddrOffset (PciAddress, Offset),
+                                  sizeof (PCI_EXPRESS_EXTENDED_CAPABILITIES_RESIZABLE_BAR_CONTROL),
+                                  &ResizableBarControl
+                                  );
+
+  DEBUG ((
+    DEBUG_INFO,
+    "OCDM: RBAR control is %X, total %u - %r\n",
+    ResizableBarControl.Uint32,
+    MIN (ResizableBarControl.Bits.ResizableBarNumber, PCI_MAX_BAR),
+    Status
+    ));
+
+  if (EFI_ERROR (Status)) {
+    return EFI_UNSUPPORTED;
+  }
+
+  ResizableBarNumber = MIN (ResizableBarControl.Bits.ResizableBarNumber, PCI_MAX_BAR);
```

Comment:
@mhaeuser Might it be worth adding a static assert here that `PCI_MAX_BAR > 0`, since the previous `ChangedBars` logic would have come into effect in that (implausible? broken?) case?

Reply:
It’s not configurable, is it? I only add STATIC_ASSERTs for non-configurable macros when the property is not obvious to hold (say inequality with another unobvious number, whether it’s a power of two, or divisible by some value). This looks fairly obvious to me, but I might be missing something.

Reply:
I thought it could sometimes be, as much as anything, a way to document an assumption made, in order for the code which follows to be correct.

But happy to accept that this is such an obvious or unavoidable 'assumption' of this coffee, that it is not worth documenting.

PS I did wonder whether the check that was removed might perhaps have been added in the first place to placate some previous static coffee analyser?

Reply:
The `ChangedBars` check which was removed because it's not needed because the thing that we are not static asserting is never going to be false!

Reply:
It took me a couple of reads, but I believe I understand you distinction for configurable vs. non-configurable constants, and when you would choose to static assert on the latter, now, tyvm.

Reply:
> The `ChangedBars` check which was removed because it's not needed because the thing that we are not static asserting is never going to be false!

Wh-What? :D The only check I suggested to "remove" was when a BOOLEAN was set to TRUE only on a single path. Instead of having a BOOLEAN, you might as well just move the conditional code there. I don't remember where or what it was. As I said, PCI is not my strength and I did not try to understand the code in any shape or form.

Reply:
> It took me a couple of reads, but I believe I understand you distinction for configurable vs. non-configurable constants, and when you would choose to static assert on the latter, now, tyvm.

Yes, just for the record, configurable constants (say PCDs or other constants that are not set in stone) are special, as they may be re-chosen with the risk of human error. So I'd ASSERT a lot more for those.